plugins/outputs/zabbix/lld.go

```diff
@@ -0,0 +1,261 @@
+package zabbix
```

Comment:
I think you can simplify this much more. What do you think of
```go
package zabbix

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"hash/fnv"
	"sort"
	"strings"
	"time"

	"github.com/influxdata/telegraf"
	"github.com/influxdata/telegraf/config"
	"github.com/influxdata/telegraf/internal"
	"github.com/influxdata/telegraf/metric"
)

const (
	lldName = "lld"
	hostTag = "host"
	empty   = `{"data":[]}`
)

type lldInfo struct {
	Hostname string
	Key      string
	Data     map[uint64]map[string]string
}

func (i *lldInfo) hash() uint64 {
	ids := make([]uint64, 0, len(i.Data))
	for id := range i.Data {
		ids = append(ids, id)
	}
	sort.Slice(ids, func(i, j int) bool { return ids[i] < ids[j] })

	h := fnv.New64a()
	h.Write([]byte(i.Hostname))
	h.Write([]byte("\n"))
	h.Write([]byte(i.Key))
	h.Write([]byte("\n"))

	// Write cannot fail
	_ = binary.Write(h, internal.HostEndianness, ids)

	return h.Sum64()
}

func (i *lldInfo) metric() (telegraf.Metric, error) {
	values := make([]map[string]string, 0, len(i.Data))
	for _, v := range i.Data {
		values = append(values, v)
	}
	data := map[string]interface{}{"data": values}
	buf, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	return metric.New(
		lldName,
		map[string]string{
			hostTag: i.Hostname,
		},
		map[string]interface{}{
			i.Key: buf,
		},
		time.Now(),
	), nil
}

type zabbixLLD struct {
	log telegraf.Logger

	// current is the collection of metrics added during the recent period
	current map[uint64]lldInfo

	// previous stores the hashes of metrics received during the previous period
	previous map[uint64]lldInfo

	// lastClear store the time of the last clear of the LLD data
	lastClear time.Time

	// clearInterval after this number of pushes, all data is considered new.
	// The idea behind this parameter is to resend known LLDs with low freq in case
	// previous sent was not processed by Zabbix.
	clearInterval config.Duration
}

// Push returns a slice of metrics to send to Zabbix with the LLD data using the accumulated info.
// The name of the metric will be always "lld" (const).
// It will have only one tag, with the host.
// It will have an uniq field, with the LLD key as the key name and the JSON data as the value
// Eg.: lld,host=hostA disk.device.fstype.mode.path="{\"data\":[...
func (zl *zabbixLLD) Push() []telegraf.Metric {
	metrics := make([]telegraf.Metric, 0, len(zl.current))
	newPrevious := make(map[uint64]lldInfo, len(zl.current))

	// Iterate over the data collected in the closing period and determine which
	// data needs to be send. This can be done by comparing the complete data
	// hash with what was previously sent (i.e. during last push). If different,
	// send a new LLD.
	seen := make(map[uint64]bool, len(zl.current))
	for _, info := range zl.current {
		// Get the hash of the complete data and remember the data for next period
		id := info.hash()
		newPrevious[id] = lldInfo{
			Hostname: info.Hostname,
			Key:      info.Key,
		}
		seen[id] = true

		// Skip already sent data
		if _, found := zl.previous[id]; found {
			continue
		}

		// Add the LLD metric
		m, err := info.metric()
		if err != nil {
			zl.log.Warnf("Marshaling to JSON LLD tags in Zabbix format: %v", err)
			continue
		}
		metrics = append(metrics, m)
	}

	// Check if we have seen the LLD in this cycle and send an empty LLD otherwise
	for id, info := range zl.previous {
		if seen[id] {
			continue
		}
		m := metric.New(
			lldName,
			map[string]string{
				hostTag: info.Hostname,
			},
			map[string]interface{}{
				info.Key: []byte(empty),
			},
			time.Now(),
		)
		metrics = append(metrics, m)
	}

	// Replace "previous" with the data of this period or clear previous
	// if enough time has passed
	if time.Since(zl.lastClear) < time.Duration(zl.clearInterval) {
		zl.previous = newPrevious
	} else {
		zl.previous = make(map[uint64]lldInfo, len(zl.previous))
		zl.lastClear = time.Now()
	}

	return metrics
}

// Add parse a metric and add it to the LLD cache.
func (zl *zabbixLLD) Add(in telegraf.Metric) error {
	// Extract all necessary information from the metric
	hostname, key, mid, data := processMetric(in)
	if hostname == "" {
		return errors.New("no hostname found")
	}

	// Try to lookup the Zabbix series in the already received metrics and
	// create a new one if not found
	series := lldSeriesID(hostname, key)
	if _, found := zl.current[series]; !found {
		zl.current[series] = lldInfo{
			Hostname: hostname,
			Key:      key,
			Data:     make(map[uint64]map[string]string),
		}
	}
	zl.current[series].Data[mid] = data

	return nil
}

func processMetric(m telegraf.Metric) (host, key string, hash uint64, tags map[string]string) {
	// Get the metric tags. The tag-list is already sorted by key name
	tagList := m.TagList()

	// Iterate over the tags and extract
	//   - the hostname contained in the host tag
	//   - the key-values of the tags WITHOUT the host tag
	//   - the LLD-key for sending the metric in the form
	//     <metric>.<tag key 1>[,<tag key 2>...,<tag key N>]
	//   - a hash for the tag the Zabbix series
	h := fnv.New64a()
	h.Write([]byte(m.Name()))
	h.Write([]byte("\n"))

	var hostname string
	keys := make([]string, 0, len(tagList))
	data := make(map[string]string, len(tagList))
	for _, tag := range tagList {
		// Extract the host key and skip everything else
		if tag.Key == hostTag {
			hostname = tag.Value
			continue
		}

		// Collect the tag keys for generating the lld-key later
		if tag.Value != "" {
			keys = append(keys, tag.Key)
		}

		// Prepare the data for lld-metric
		dataKey := "{#" + strings.ToUpper(tag.Key) + "}"
		data[dataKey] = tag.Value

		// Hash the metric series
		h.Write([]byte(tag.Key))
		h.Write([]byte("\n"))
		h.Write([]byte(tag.Value))
		h.Write([]byte("\n"))
	}

	var lldKey string
	if len(keys) > 0 {
		lldKey = m.Name() + "." + strings.Join(keys, ".")
	}

	return hostname, lldKey, h.Sum64(), data
}

func lldSeriesID(hostname, key string) uint64 {
	h := fnv.New64a()
	h.Write([]byte(hostname))
	h.Write([]byte("\n"))
	h.Write([]byte(key))
	h.Write([]byte("\n"))
	return h.Sum64()
}

```

Reply:
@srebhan I have partially adapted the current LLD tests to this implementation but I think it is missing the part that cleans `zl.current`.

The tests trying to reproduce how an empty LLD is sent are failing.

Reply:
@srebhan this part of the code
```
func (i *lldInfo) metric() (telegraf.Metric, error) {
	values := make([]map[string]string, 0, len(i.Data))
	for _, v := range i.Data {
		values = append(values, v)
	}
	data := map[string]interface{}{"data": values}
	buf, err := json.Marshal(data)
```
generates the string that should be sent to Zabbix. Eg.: 
```json
{"data": [{"{#FOO}": "1", "{#BAR}": "2"}, {"{#FOO}": "3", "{#BAR}": "4"}]}
```

As the access to i.Data [is random](https://go.dev/blog/maps#TOC_7.:~:text=iteration%20order%20is%20not%20specified), that string could be different in each execution, so it fails in the tests when compared with a manually created metric (https://github.com/influxdata/telegraf/pull/13739/files#diff-107b0e2d6f7bc2b5e8a9a1f09c7d1563047ea92446e5cef98ea80bae77d1dfa3R445)

One solution could be to [sort using the key](https://stackoverflow.com/questions/23330781/collect-values-in-order-each-containing-a-map), but looks like a waste of resources, as Zabbix does not need it.

¿Any ideas?

Maybe a custom ``testutil.RequireMetricsEqual`` that could ignore the order of that `data` array.

Or maybye the test is not well oriented...

Reply:
@srebhan I have tried to use this implementation, but it does not work for a particular case. I have created another branch with this implementation and migrating the rest of code and tests here: https://github.com/datadope-io/telegraf/commit/7a6c87ede49c7fee948fdf9e41fd47967c6e9aa4

In the commit I explain which is the problem.